l10n_de_tax_statement/tests/test_l10n_de_tax_statement.py

```diff
@@ -0,0 +1,356 @@
+        self.statement_1.statement_update()
+        for line in self.statement_1.line_ids:
+            self.assertTrue(line.view_base_lines())
+            self.assertTrue(line.view_tax_lines())
+
+        # first post
+        self.statement_1.statement_update()
+        self.statement_1.post()
+        self.assertEqual(self.statement_1.state, 'posted')
+        self.assertTrue(self.statement_1.date_posted)
+
+        for line in self.statement_1.line_ids:
+            self.assertTrue(line.view_base_lines())
+            self.assertTrue(line.view_tax_lines())
+
+        # then finalize
+        self.statement_1.finalize()
+        self.assertEqual(self.statement_1.state, 'final')
+        self.assertTrue(self.statement_1.date_posted)
+
+        with self.assertRaises(UserError):
+            self.statement_1.write({'name': 'Test Name Modified'})
+        with self.assertRaises(UserError):
+            self.statement_1.write({'state': 'posted'})
+        with self.assertRaises(UserError):
+            self.statement_1.write({'date_posted': fields.Datetime.now()})
+        with self.assertRaises(UserError):
+            self.statement_1.unlink()
+        for line in self.statement_1.line_ids:
+            self.assertTrue(line.view_base_lines())
+            self.assertTrue(line.view_tax_lines())
+            with self.assertRaises(UserError):
+                line.unlink()
+
+        self.assertEqual(self.statement_1.tax_total, 0.)
+
+    def test_03_reset(self):
+        self.statement_1.reset()
+        self.assertEqual(self.statement_1.state, 'draft')
+        self.assertFalse(self.statement_1.date_posted)
+
+        self.assertEqual(self.statement_1.tax_total, 0.)
+
+        self.statement_1.statement_update()
+        for line in self.statement_1.line_ids:
+            self.assertTrue(line.view_base_lines())
+            self.assertTrue(line.view_tax_lines())
+
+    def test_04_write(self):
+        self.statement_1.post()
+        with self.assertRaises(UserError):
+            self.statement_1.write({'name': 'Test Name'})
+
+        self.assertEqual(self.statement_1.tax_total, 0.)
+
+    def test_05_unlink_exception(self):
+        self.statement_1.post()
+        with self.assertRaises(UserError):
+            self.statement_1.unlink()
+
+    def test_06_unlink_working(self):
+        self.statement_1.unlink()
+
+    def test_07_update_exception1(self):
+        self.statement_1.post()
+        with self.assertRaises(UserError):
+            self.statement_1.statement_update()
+
+    def test_08_update_exception2(self):
+        self.config.unlink()
+        with self.assertRaises(UserError):
+            self.statement_1.statement_update()
+
+    def test_09_update_working(self):
+        self.invoice_1._onchange_invoice_line_ids()
+        self.invoice_1.action_invoice_open()
+        self.statement_1.statement_update()
+        self.assertEqual(len(self.statement_1.line_ids.ids), 22)
```

Comment:
Aren't there 47 tags?

Reply:
yes you are right, 22 comes from the NL module.